### ISSUE TITLE
Implementing Benchmark Simulation Model No. 2

### ISSUE THREAD

**Issue by adam-a-a** [watertap-org/watertap#934]
### Description

Our goal is to create a flowsheet for the Benchmark Simulation Model No. 2 (BSM2) which is a standard used in the wastewater treatment industry. BSM2 represents a full wastewater treatment plant (WWTP) that includes the activated sludge process as well as an anaerobic digestor, along with other units. 

**The ultimate goal is to utilize this detailed, plant-scale model, and prioritized BSM2-related model enhancements that would be required, to facilitate detailed technology comparisons via cost optimization for innovations proposed by teams under the IEDO 2336 effort.**

### Motivation

This will provide a detailed baseline for wastewater treatment processes and could be used for cost optimization studies.

### Running TO-DO:
- [ ] #985 
- [ ] Create a modified ASM2D that would be compatible with precipitation PCM extension for ADM1
- [ ] #1318 
- [ ] Add TSS, COD, BOD5, Kjeldahl nitrogen to ADM1
- [ ] Add TSS, ISS, and/or VSS calculations directly to ASM/ADM models that are missing them and phase out ASM config option on thickener/dewaterer  (**started with modified ASM2D and modified ADM1 in #1078 but needs additions on ASM1, ASM2d and ADM1**)
- [ ] Review pH refinements in ASM and ADM models and incorporate into modified versions
- [ ] Secondary clarifier unit model - Takács model
- [ ] primary clarifier unit model - BSM implementation of model

### Questions , points of discussion, and minor issues
- [ ] Implement variation of ASM1 --> ADM1 translator block that includes **modified** Copp Interface
- [ ] storage tank unit model?
- [ ] Modify auxiliary units to be compatible with other ASM models #1078 - does primary clarifier in true implementation need this?

- [ ] Address simplification in modified ASM2d where S_IC does not account for carbonate and dissolved CO2 (Serralta 2004)
- [ ] make a baseline activated sludge reactor unit model?
- [ ] Track hydrogen sulfide (H2S) in ADM such that `Z_h2s` != 0?
- [ ] storage tank unit model?

### Tasks complete by Sept 2025
- [x] Delete simple modified ASM2d model? #1531 
- [x] Add TSS, COD, BOD5, Kjeldahl nitrogen to ASM2d #1492
- [x] Refinement of scaling in ASM2D #1533 
- [x] Track aggregate metrics throughout BSM2 flowsheet (pH, P components, N components) - see Solon 2017 #1244 

### Tasks completed by June 2024 release:
- [x] Export "BSM2" w/asm2d/adm1 to gui
- [x] Replace Expr_ifs in ASM2d/ADM1 translator with smooth_max #1362 

### Tasks completed by Mar 2024 release:
- [x] Full BSM2 flowsheet using ADM1 & ASM2d #1113 , #1276
- [x] Add aeration energy to ASM flowsheets (added to BSM2 flowsheet) #1244 

### Tasks completed by Dec 2023 release:
- [x] costing parameters and expressions for each unit in BSM2 #1217 
- [x] Add TSS, COD, BOD5, Kjeldahl nitrogen to ASM1- #1243 

### Tasks completed by Sept 2023 release:
- [x] Export BSM2 (with ASM1/ADM1) to GUI #1125 
- [x] Modify dewaterer and thickener to be compatible with other ASM models #1078
- [x] Full BSM2 flowsheet using ADM1 & ASM1 #1038 
- [x] Implement translator block for ASM2D-> modified ADM1 w/ P extension #1074 
- [x] Update ADM1 & modified ADM1 documentation based on the refinements in #1038 
- [x] Refinement of initial ADM1/AD implementation, including scaling/initialization as needed #909 
- [x] Refinement of scaling in ASM1 #1038 
- [x] #1002 

### Tasks completed by the June 2023 release:
- [x] ElectroN-P sub-flowsheet w/ AD -> ADM1/ASM2d -> ElectroN-P #1032 
- [x] Create a new ASM2d model that does not make as many simplifications #1031 #1036
- [x] Create ADM1/ASM2d translator using the new ASM2d model #1058 #1063 
- [x] #1013 
- [x] #1017 
- [x] Implement translator block for ADM1-> ASM1 #933 
- [x] Implement translator block for ASM1-> ADM1 #962 
- [x] Documentation for ASM1-->ADM1 for Copp Interface #1000 
- [x] Create a new ADM1 model that accounts for phosphorus #948 
- [x] Implement translator block for ADM1-> ASM2D #935 (waiting on new ADM1 model #948)
- [x] Documentation for ADM1 w/ P extension and/or precipitation PCM extension #996 
- [x] Dewatering unit model #1010 
- [x] Sludge thickening unit model #1010 
- [x] Documentation for dewatering unit and thickener #1017 
- [x] #1020 
- [x] Remove `s_ic_rxns` for loop in `adm1_reactions`? - no longer relevant/delete
- [x] #941 Addressed in PR #997
- [x] Address question about CO2 only in gas phase for ADM1 #1004 #1012

**Comment by lbianchi-lbl** [comment]
@adam-a-a @agarciadiego @luohezhiming confirm that it would be good (and doable) to have this done for June.

**Comment by adam-a-a** [comment]
To elaborate, we think it's be doable to have an initial implementation of the full flowsheet functioning with much of the unit documentation in place. However, some of the additional items around model enhancement may spill over past June.

**Comment by adam-a-a** [comment]
Noting this down as I work on #1078 : noticed that the thickener doesn't have its own `calculate_scaling_factors` nor `initialize` methods. Instead it would use said methods from the IDAES separator unit, which the thickener class inherits from. Furthermore, testing of scaling is not split into its own test; it is packed in with the testing of initialization of the thickener, and no exploration on badly_scaled_variables or unscaled_constraints is conducted.

**Comment by adam-a-a** [comment]
This relates to #1318  as well
